### PR TITLE
refactor: Refactor media to support multiple calls

### PR DESCRIPTION
**Pull request by bzang**

### CODE REVIEW

File: packages/node_modules/@ciscospark/redux-module-media/src/actions.js [hunk 1]
```diff
@@ -123,39 +78,29 @@ export function updateWebRTCSupport(supported) {
...
-  call.on(`remoteVideoMuted:change`, () => dispatch(updateCallState(call)));
-}
-
-function bindMediaActions(dispatch, call) {
-  call.on(`remoteMediaStream:change`, () => dispatch(updateRemoteMedia(call)));
-  call.on(`localMediaStream:change`, () => dispatch(updateLocalMedia(call)));
-}
 
-export function callUser({userIdOrEmail, constraints = {audio: true, video: true}}, spark) {
+export function callUser({
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/redux-module-media/src/actions.js]: Can we get jsdoc on this??

File: packages/node_modules/@ciscospark/redux-module-media/src/reducer.js [hunk 2]
```diff
@@ -3,142 +3,81 @@ import {fromJS} from 'immutable';
...
-  startTime: null,
-  incomingCall: {
-    hasAcknowledged: false,
-    call: null
-  },
-  callState: {
-    sendingAudio: null,
-    sendingVideo: null
-  },
+  calls: {},
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/redux-module-media/src/reducer.js]: Do we care what order we receive these in? Should this be an ordered map?

Comment (bzang) [inline packages/node_modules/@ciscospark/redux-module-media/src/reducer.js]: Order they come in really doesn't matter. In recents it's the order of last seenable activity that matters.

**Comment by bzang** [comment]
Ready for a final review @adamweeks

File: packages/node_modules/@ciscospark/redux-module-media/src/actions.js [hunk 3]
```diff
@@ -29,91 +28,47 @@ export function updateStatus(status) {
...
-  return {
-    type: UPDATE_REMOTE_MEDIA,
-    payload: {
-      call
-    }
-  };
-}
-
-export const CHECKING_WEBRTC_SUPPORT = `media/CHECKING_WEBRTC_SUPPORT`;
 export function checkingWebRTCSupport() {
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/redux-module-media/src/actions.js]: Should we match the function name to the action name?

File: packages/node_modules/@ciscospark/redux-module-media/src/actions.js [hunk 4]
```diff
@@ -175,33 +128,34 @@ export function hangupCall(call) {
...
           // Disconnects will happen if the remote party hangs up before we accept/decline
           incomingCall.once(`disconnected`, removeIncoming);
           // Once we are connected, our connected call state will handle disconnects
           incomingCall.once(`connected`, () => incomingCall.off(`disconnected`, removeIncoming));
-          bindCallActions(dispatch, incomingCall);
-          bindMediaActions(dispatch, incomingCall);
+          bindEvents(dispatch, incomingCall);
         });
     });
     dispatch(updateStatus({isListening: true}));
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/redux-module-media/src/actions.js]: Needs a return here.

File: packages/node_modules/@ciscospark/redux-module-media/src/actions.js [hunk 5]
```diff
@@ -219,7 +173,7 @@ export function declineIncomingCall(incomingCall) {
   return (dispatch) => {
     incomingCall.reject()
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/redux-module-media/src/actions.js]: Return needed here and on the dispatch in the then

File: packages/node_modules/@ciscospark/redux-module-media/src/actions.js [hunk 6]
```diff
@@ -228,13 +182,14 @@ export function acceptIncomingCall(incomingCall) {
   return (dispatch) => {
     incomingCall.answer({offerOptions: {offerToReceiveVideo: true, offerToReceiveAudio: true}})
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/redux-module-media/src/actions.js]: `return` 😄

File: packages/node_modules/@ciscospark/redux-module-media/src/actions.js [hunk 7]
```diff
@@ -228,13 +182,14 @@ export function acceptIncomingCall(incomingCall) {
...
   };
 }
 
+
 /**
- * Check for brwoser webRTC support
+ * Check for browser webRTC support
  *
  *  @param {object} spark
  *  @returns {Promise}
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/redux-module-media/src/actions.js]: Can we `return` here too?

File: packages/node_modules/@ciscospark/redux-module-media/src/actions.js [hunk 8]
```diff
@@ -247,3 +202,47 @@ export function checkWebRTCSupport(spark) {
...
+
+
+function constructCallState(call) {
+  let remoteAudioMuted, remoteVideoMuted;
+  try {
+    remoteAudioMuted = call.remoteAudioMuted;
+    remoteVideoMuted = call.remoteVideoMuted;
+  }
+  catch (e) {
+    // eslint-disable-line no-console
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/redux-module-media/src/actions.js]: There's no console to `no-console` here 😛

File: packages/node_modules/@ciscospark/redux-module-media/src/actions.js [hunk 9]
```diff
@@ -247,3 +202,47 @@ export function checkWebRTCSupport(spark) {
...
+  }
+  return {
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/redux-module-media/src/actions.js]: I like moving this here and out of the reducer!

File: packages/node_modules/@ciscospark/widget-meet/src/container.js [hunk 10]
```diff
@@ -46,33 +46,61 @@ export class MeetWidget extends Component {
...
-      if (startCall
-        && !media.getIn([`status`, `isDialing`])
-        && !media.getIn([`status`, `isConnected`])) {
-        this.handleCall();
-      }
+    if (nextProps.media.getIn([`webRTC`, `isSupported`])) {
       this.setup(nextProps);
     }
+
+    const currentUser = users.get(`currentUser`);
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/widget-meet/src/container.js]: Do we want to do any of this if webRTC is not supported?

Comment (bzang) [inline packages/node_modules/@ciscospark/widget-meet/src/container.js]: good call

Comment (adamweeks) [inline packages/node_modules/@ciscospark/widget-meet/src/container.js]: What were we going to do here?

File: packages/node_modules/@ciscospark/widget-meet/src/container.js [hunk 11]
```diff
@@ -234,105 +258,122 @@ export class MeetWidget extends Component {
...
-      username: incomingCall.from.person.name,
+      username: fromPersonName,
       message: formatMessage(messages.incomingCallMessage),
       avatar: avatarImage
     };
     this.props.createNotification(incomingCall.locusUrl, NOTIFICATION_TYPE_OTHER, details);
   }
 
-  render() {
+  render() { // eslint-disable-line complexity
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/widget-meet/src/container.js]: We might want to break these up into smaller functions at some point so it isn't as complex.

**Pull request merged**